### PR TITLE
CU-1ux3e25 CU-1q4x2t6 CU-1pdtjrz finalizing lending features

### PR DESCRIPTION
**Pull request by dzmitry-lahoda**
- stake during market creation
- rent for borrow and return back on remove
- improving docs
- will add price inverse in next pr to test
- checked BF if can be used to reward debt holders

Signed-off-by: Dzmitry Lahoda <dzmitry@lahoda.pro>

### CODE REVIEW

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-1pdtjrz lending: for each liquidation give some pica (Min - Weight for call - NoPay Dispatch Return))](https://app.clickup.com/t/1pdtjrz)

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-1ux3e25 lending: ensure lending token can be used for staking](https://app.clickup.com/t/1ux3e25)

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-1q4x2t6 lending: ddos: take rent per pool](https://app.clickup.com/t/1q4x2t6)

File: frame/bonded-finance/src/tests.rs [hunk 1]
```diff
@@ -179,6 +179,42 @@ prop_compose! {
 			  }
 }
 
+#[test]
+fn bond_offer_cc_e005bce405358673d3249e53d51e474d4b135bdf8f2b0b82f94f525c0efbdbf8() {
```
Comment (hussein-aitlahcen) [inline frame/bonded-finance/src/tests.rs:183]: If I remember correctly, proptest will automatically replay the tests saved in the `proptest-regressions` directory, we should create a unit test if the test was a corner case, but in this case it's just precision (not really something different in the logic)

File: frame/composable-support/src/validation.rs [hunk 2]
```diff
@@ -53,16 +56,50 @@ impl<T: Validate<U> + Validate<V>, U, V> Validate<(U, V)> for T {
...
+// so if we will need more than 3, can consider it
+impl<T: Validate<U> + Validate<V> + Validate<W>, U, V, W> Validate<(U, V, W)> for T {
+	#[inline(always)]
+	fn validate(self) -> Result<Self, &'static str> {
+		let value = Validate::<U>::validate(self)?;
+		let value = Validate::<V>::validate(value)?;
+		let value = Validate::<W>::validate(value)?;
+		Ok(value)
+	}
+}
```
Comment (hussein-aitlahcen) [inline frame/composable-support/src/validation.rs:70]: This is already derived if you use `(U, (V, (W, Valid))`

File: frame/lending/src/lib.rs [hunk 3]
```diff
@@ -216,6 +209,36 @@ pub mod pallet {
...
+		/// not exactly platform rent only.
+		///
+		/// # Why borrow amount priced by Oracle?
+		///
+		/// We depend on Oracle to price in Lending. So we know price anyway.
+		/// We normalized price over all markets and protect from spam all possible pairs equally.
+		/// Locking borrow amount ensures manager can create market wit borrow assets, and we force
+		/// him to really create it.
+		#[pallet::constant]
+		type MarketCreationStake: Get<Self::Balance>;
```
Comment (hussein-aitlahcen) [inline frame/lending/src/lib.rs:234]: Nice explanation!

File: frame/composable-support/src/validation.rs (hunk 2 above)
Comment (dzmitry-lahoda) [reply frame/composable-support/src/validation.rs:70]: I see. But with this user has 2 options for convenience? :)

File: frame/bonded-finance/src/tests.rs (hunk 1 above)
Comment (dzmitry-lahoda) [reply frame/bonded-finance/src/tests.rs:183]: that is weird. i created file by rules. but it was not triggered. strange.

Comment (hussein-aitlahcen) [reply frame/bonded-finance/src/tests.rs:183]: Weird because it's stated in their doc https://altsysrq.github.io/proptest-book/proptest/failure-persistence.html

File: frame/lending/src/lib.rs [hunk 4]
```diff
@@ -1081,6 +1146,30 @@ pub mod pallet {
...
+					initial_pool_size,
+					false,
+				)?;
+
+				// TODO: discuss on why we do not reposit all amount to vault
+				// <T::Vault as Vault>::deposit(
+				// 	&borrow_asset_vault,
+				// 	&Self::account_id(&market_id),
+				// 	initial_pool_size,
+				// )?;
```
[the remark this answers is not in the document]
Comment (dzmitry-lahoda) [reply frame/lending/src/lib.rs:1171]: if consider it as normal lend, she can take easy back as it is not locked - spam. i consider this as irreversible rent. may be can set minimal amount into vault which can never return? but overall ok of vault will take it back. it just spam protection. i have seen 500 USD and 10K, depending on market. we can check with product later

File: frame/lending/src/lib.rs [hunk 5]
```diff
@@ -1196,16 +1296,25 @@ pub mod pallet {
...
+						burn_amount,
+						true,
+					)
+					.expect("can always release held debt balance");
+					<T as Config>::MultiCurrency::burn_from(
+						debt_asset_id,
+						beneficiary,
+						burn_amount,
+					)
+					.expect("can always burn debt balance");
```
[the remark this answers is not in the document]
Comment (dzmitry-lahoda) [reply frame/lending/src/lib.rs:1314]: will check

File: frame/lending/src/lib.rs [hunk 6]
```diff
@@ -1363,15 +1486,18 @@ pub mod pallet {
...
 			ensure!(
-				<T as Config>::Currency::can_deposit(market.collateral, &market_account, amount) ==
-					DepositConsequence::Success,
+				<T as Config>::MultiCurrency::can_deposit(
+					market.collateral,
+					&market_account,
+					amount
+				) == DepositConsequence::Success,
 				Error::<T>::TransferFailed
 			);
```
[the remark this answers is not in the document]
Comment (dzmitry-lahoda) [reply frame/lending/src/lib.rs:1502]: will clean up checks in next pr as optimization (and will check that each caller is transactional (and will document such).

File: frame/lending/src/lib.rs [hunk 7]
```diff
@@ -1277,64 +1286,56 @@ pub mod pallet {
...
-					)
-					.expect("can always release held debt balance");
-					<T as Config>::MultiCurrency::burn_from(
-						debt_asset_id,
-						beneficiary,
-						total_repay_amount,
-					)
-					.expect("can always burn debt balance");
+
+				// BUG: so each time we repay, we must burn from market and from account, evidently
```
Comment (dzmitry-lahoda) [inline frame/lending/src/lib.rs:1290]: @hussein-aitlahcen this bug:)

**Comment by dzmitry-lahoda** [comment]
@seunlanlege pallet was removed from all runtimes, so need to re add it again to have any benchmarks. there is task about it. i wrote about it in common channel too.

**Comment by seunlanlege** [comment]
looks like simnode passed, congrats :D, just need to fix cargo fmt

**Review by hussein-aitlahcen — APPROVED** [review]

**Review by seunlanlege — APPROVED** [review]
